### ISSUE TITLE
cadl compile failed when package.json doesn't include all emitters defined in cadl-project.yaml

### ISSUE THREAD

**Issue by dw511214992** [microsoft/cadl#1457]
## How to reproduce
1. Get [my cadl project](https://github.com/Azure/azure-rest-api-specs/tree/dw/cadl-bug2/specification/cognitiveservices/OpenAI.Inference)
2. `npm install`
3. `npx cadl compile . --emit @azure-tools/cadl-java --arg "java-sdk-folder=/tmp/java"`

## Result
Encounter Error:
```
error import-not-found: Couldn't resolve import "@azure-tools/cadl-python"
error emitter-not-found: Emitter with name @azure-tools/cadl-python is not found.

Found 2 errors.
```

## Expect
Generate java sdk succesfully.

## Root cause analysis:
The `package.json`'s dependencies doesn't include emitter `@azure-tools/cadl-python`, which is defined in `cadl-project.yaml`. 
As discussed with @weshaggard  and @allenjzhang , `package.json` doesn't include any sdk emitters by default. When running in sdk automation pipeline, we only install the target sdk emitter in `package.json`. For example, I want to generate java sdk and I have installed `@azure-tools/cadl-java` in `package.json`, and then run command `npx cadl compile . --emit @azure-tools/cadl-java --arg "java-sdk-folder=/tmp/java"`, cadl should generate java sdk and don't check whether `@azure-tools/cadl-python` exist or not when I specify the java emitter in command.
I also test this issue in cadl v0.37.* and it works well. So this issue only exists in v0.38.*.

**Comment by dw511214992** [comment]
@timotheeguerin

**Comment by allenjzhang** [comment]
Ack. @timotheeguerin, in order for this to work, we would need to skip loading verification of libraries that is not listed or being emitted.

**Comment by raych1** [comment]
@dw511214992 validated this issue was introduced from v0.38.*

**Comment by timotheeguerin** [comment]
This was done by design here https://github.com/microsoft/cadl/issues/1211. Adding @m-nash and @tjprescott which both got frustrated when this wasn't being validated and had a wrong emitter name.

Now we have to weight the 2 side of this issue and I see a few options:
1. Only do the validation on the `--option` flag not in the config
   - Pro: We still get to fail the cli use
   - Cons: We are now inconsistent + not sure if we can easily do that as the validation is later in the process
2. Provide a flag `strict-emitter` which can be turned on/off(Can decide on the default)
   - Pro: This could give an escape hatch for azure sdk repo
   - Cons: Adds a new flag

Thing is this scenario feels very specific to azure specs repo flow. With option 1 we are making it worse for all user of Cadl to satisfy a workaround. I think option 2 with a default of `strict-emitter: true` would be better but also this brings a flag that might become obsolete for our use case if we ever get to be able to import urls.

**Comment by timotheeguerin** [comment]
# Proposal 

## Add a new cadl config property

```cadl
model CadlConfig {
  "validate-emitter-options"?:
    | "strict"          // Validate everything name & options
    | "installed-only"  // Skip uninstalled emitters - (Do we need this one?)
    | "emit-only"       // Only validate options for emitter running
    | "off"             // Skip uninstalled emitters and don't validate the options
    = "strict";         // default to "strict"
}
```

```yaml
emit:
  - cadl-autorest
options:
  cadl-autorest:
    opt1: "abc"
    unknown-opt: "abc"

  cadl-java:
    opt1: "abc"
    unknown-opt: "abc"

  cadl-python:
    opt1: "abc"
    unknown-opt: "abc"
```

```json
{
  "dependencies": {
    "cadl-autorest": "latest",
    "cadl-java": "latest"
  }
}
```

## `strict` - Validate everything name & options

- complain `cadl-java` is not installed
- complain `cadl-csharp` is not installed
- complain `cadl-autorest.unknown-opt` doesn't exist
- complain `cadl-java.unknown-opt` doesn't exist (even though it is not run)

## `installed-only` - Skip uninstalled emitters - (Do we need this one?)

- complain `cadl-autorest.unknown-opt` doesn't exist
- complain `cadl-java.unknown-opt` doesn't exist (even though it is not run)

## `emit-only` - Only validate options for emitter running

- complain `cadl-autorest.unknown-opt` doesn't exist

## `off` - Skip uninstalled emitters and don't validate the options

No errors


## Alternative

If we don't want a new flag then the only solution is to really just disable validation and act like `emit-only` from above.

**Comment by timotheeguerin** [comment]
Updated the error to be a warning to mitigate the issue #1460 until resolution of above design.

**Comment by tjprescott** [comment]
I ran into this as well and it annoyed me. The "fix" was to comment out the emitter options for the offending emitter. However, since I wasn't trying to emit that particular language, I was surprised it even checked those options. But given the details @timotheeguerin shared, I'm thinking the warning may be the best option. The problem before was that it just silently failed and you had no idea anything was wrong (except that it didn't do what you expected).

My vote would probably be for no new flag and just emit the warning instead of error.

**Comment by dw511214992** [comment]
Thanks @timotheeguerin  I like this solution which only adds a warning. Do you want to close this issue? Or let's keep it open if you want to figure out a better solution?

**Comment by timotheeguerin** [comment]
Yeah lets keep it open, my guess is that a warning is not going to be working if you want to prevent specs from being merged with warnings in the future.

**Comment by markcowl** [comment]
- the error was made a warning to unblock this usage
- design discussion needed for proposal

**Comment by m-nash** [comment]
Currently dotnet doesn't run into this issue because we are replacing the package.json file.  We then run a [similar command](https://github.com/Azure/azure-sdk-for-net/blob/main/eng/scripts/Cadl-Project-Generate.ps1#L58) to the one above and it works just fine.

Unless some new issue was introduced over the holidays that I am not aware of.

**Comment by weshaggard** [comment]
Just to add my 2-cents I like option "emit-only" as the default and no warning or errors for options that are not used in the current context. We don't want to have normal cases where we output warnings as that will simply turn into ignored noise and likely not help anyone. 

@m-nash I'm guessing .NET will hit this as soon as you start consuming cadl-project.yaml files that contains options for all the emitters.

**Comment by m-nash** [comment]
@weshaggard ah yes you are correct I just tried it and I guess up until now its been single emitter in the cadl-project.yaml.

**Comment by timotheeguerin** [comment]
@weshaggard I agree that for the specs repo `emit-only` makes the most sense to use as the default but for anybody else using Cadl they will most likely have a normal setup where they have the whole picture of their emitters and in this case errors won't show up until they ask for it(Which most people won't do) and then start getting confused why things don't work when they have a typo.

My opinion is to stick to `strict` as the Cadl default but then we have the option set to `emit-only` for each `cadl-project.yaml` in the specs repo

**Comment by weshaggard** [comment]
I'm fine with having this be an option to control the behavior as well. I just don't like use skipping the option and always outputting a warning.

**Comment by tjprescott** [comment]
> We don't want to have normal cases where we output warnings as that will simply turn into ignored noise and likely not help anyone.

It isn't a normal case where you output a warning. Ignored noise seems like the least disruptive solution here to me.

**Comment by m-nash** [comment]
I am not sure why a new option is needed.  I think the intent of the user is quite clear when they only add one emitter to the command line.  Seems odd to add another option that means "Yes I really did mean it when I typed that other option".

I would suggest that cadl compile validates whatever is in the cadl-project.yaml unless an `--emit` option was passed in then it only validates that one.

Unless I am missing something here.

**Comment by timotheeguerin** [comment]
@m-nash The problem is if you have lets say `--emit my-emitter` but in the options you use is like that lets say

```yaml
options:
  myemitter: # <-- here I have a typo
     ... 
```

Now this doesn't complain because we are only validating options for `my-emitter` and there is none here, so now its not reading options and you don't know why, you got into this exact problem a while back.

**Comment by m-nash** [comment]
@timotheeguerin I think this is fine.  Its perfectly valid for someone to want to use all default options for an emitter so finding a match isn't always guaranteed.

**Comment by timotheeguerin** [comment]
Yeah but here they don’t want the default option they have configured them but with a typo.

**Comment by m-nash** [comment]
If you can determine it was a typo, its probably safe to throw a warning saying "We think you made a typo", but I would still say the second flag isn't needed.

**Comment by tjprescott** [comment]
My vote would be no flag, validate everything and throw warnings.

**Comment by weshaggard** [comment]
I also think there is no real way to find all the typos but hopefully they can eventually figure out that the options they think they are setting aren't being used. I would be fine with validating all if you don't pass the --emit flag but if you do pass that flag I would suggest only validating that one emitter set of options. 

I definitely don't think it is a good idea to have warnings always outputted when someone uses a single emitter and have other emitter options in their config file. We cannot have cases like this outputting warnings. While it might not seem like a big deal for local usage it is a huge deal when using this in pipelines where you are trying to keep warning and error clean. In a lot of scenarios, we treat warnings as errors to make sure we stay clean.

**Comment by tjprescott** [comment]
> I definitely don't think it is a good idea to have warnings always outputted when someone uses a single emitter and have other emitter options in their config file. 

But that's the thing--that doesn't happen as long as the other emitter options are valid.

**Comment by weshaggard** [comment]
>But that's the thing--that doesn't happen as long as the other emitter options are valid. 

Perhaps I've misunderstood but my understanding is that if you pass --emit option to run only one emitter and there are other emitter options in the cadl-project.yaml it will warn about those because it doesn't see or find those emitters.

**Comment by tjprescott** [comment]
No, that's only the case if the emitter in question is not installed. If it is installed, then you won't get any warnings whether you use `--emit` or not. The issue here was that the poster removed an emitter from package.json so it was not installed but did not remove the emitter options.

**Comment by weshaggard** [comment]
@tjprescott I believe the plan is to not have all the emitters in the package.json file in the specs repo to avoid conflicts and other issues but we do plan to have them all in the cadl-project.yaml. See the sample in https://github.com/Azure/azure-rest-api-specs/pull/21722. That means our normal case in the spec repo will hit this case.

**Comment by tjprescott** [comment]
> I believe the plan is to not have all the emitters in the package.json file in the specs repo to avoid conflicts and other issues but we do plan to have them all in the cadl-project.yaml.

Then yes, this would be problematic. Combined with @m-nash's comment:
> dotnet doesn't run into this issue because we are replacing the package.json file. 

It sounds like there is a larger, more pressing issue that needs to be created and discussed rather than warnings and flags in the CLI.

**Comment by tjprescott** [comment]
Speaking with @m-nash, it does seem like we should change this. Based on our discussions, it seems like the proposal alternative is preferred:
> If we don't want a new flag then the only solution is to really just disable validation and act like emit-only from above.

**Comment by markcowl** [comment]
I think we are agreed that 'installed-only' seems unuseful, so the things to decide are:

Proposal 1:
- Have 'strict', 'emit-only', 'off' setting for validation
- Decide on what is the default in this case

Proposal 2:
- Just validate the options for the emitter being used

Note that there is a separate design issue for adding library options, and we want to make sure any proposal is consistent with that one: https://github.com/microsoft/cadl/issues/1425

**Comment by markcowl** [comment]
# Revised Proposal 

## Add a new cadl config property

```cadl
model CadlConfig {
  "validate-emitter-options"?:
    | "strict"          // Validate everything name & options
    | "emit-only"       // Only validate options for emitter running
    | "off"             // Skip uninstalled emitters and don't validate the options
    = "strict";         // default to "strict"
}
```

```yaml
emit:
  - cadl-autorest
options:
  cadl-autorest:
    opt1: "abc"
    unknown-opt: "abc"

  cadl-java:
    opt1: "abc"
    unknown-opt: "abc"

  cadl-python:
    opt1: "abc"
    unknown-opt: "abc"
```

```json
{
  "dependencies": {
    "cadl-autorest": "latest",
    "cadl-java": "latest"
  }
}
```

## `strict` - Validate everything name & options

- complain `cadl-java` is not installed
- complain `cadl-csharp` is not installed
- complain `cadl-autorest.unknown-opt` doesn't exist
- complain `cadl-java.unknown-opt` doesn't exist (even though it is not run)

## `emit-only` - Only validate options for emitter running

- complain `cadl-autorest.unknown-opt` doesn't exist

## `off` - Skip uninstalled emitters and don't validate the options

No errors


## Scaffolding Templates
- Update the data-plane scaffolding templates to specify the `emit-only` option.

**Comment by markcowl** [comment]
end result: switch to 'lazy' evaluation and do not provide an additional option.
